Preparing for Testnet Deployment

tegefaulkes opened the pull request:
### Description
Note that this has been migrated from MatrixAI/Polykey#326.

<!-- Write your description about what this PR is about. -->
Once MatrixAI/Polykey#310 has been merged, we'll finally be able to move onto the deployment of our testnet into AWS.

I foresee this to be achievable in a few stages:
1. Deploy Polykey onto AWS and ensure we have a working remote keynode (#194)
2. Create the testnet seed nodes and store these securely somewhere (#285)
3. Use these seed nodes for our domain name (#177)

* Relates to MatrixAI/Polykey#159 
* Relates to MatrixAI/Polykey#161 
* Migrated from MatrixAI/Polykey#326 
* Relates to MatrixAI/Polykey#297 

### Issues Fixed

* ~~Fixes MatrixAI/Polykey#324~~ being completed in MatrixAI/Polykey#310

Testnet deployment:

* Relates to MatrixAI/Polykey#194
* Relates to MatrixAI/Polykey#285 
* Relates to MatrixAI/Polykey#177 

NodeGraph Structure:

* Relates MatrixAI/Polykey#244 
* Fixes MatrixAI/Polykey#317 
* Fixes MatrixAI/Polykey#158
* Fixes MatrixAI/Polykey#359 

Node Adding Policies:

* Fixes MatrixAI/Polykey#212 
* Fixes MatrixAI/Polykey#322
* Fixes MatrixAI/Polykey#344 
* Fixes MatrixAI/Polykey#363 

Node Removal Policies:

* Fixes MatrixAI/Polykey#150 
* Fixes MatrixAI/Polykey#226 
* Fixes MatrixAI/Polykey#345

### Tasks

- ~~[ ] 2. Complete MatrixAI/Polykey#324~~ - being completed in MatrixAI/Polykey#310
- [x] 1. Read through and revisit all previous issues
- ~~[ ] 3.  Complete MatrixAI/Polykey#194~~ - Extracted out, should be finished after this PR merges
  * Ultimately provides us with a deployment of Polykey on AWS!
  * Task list there should be sufficient to follow process.
- ~~[ ] 4. Complete MatrixAI/Polykey#285~~ - Extracted out, should be finished after this PR merges
- ~~[ ] 5. Complete MatrixAI/Polykey#177~~ - Extracted out, should be finished after this PR merges
- ~~[ ] 6. Created automated testing that utilises the testnet~~ - to be done in MatrixAI/Polykey#357 
  - ~~[ ] - MatrixAI/Polykey#159~~
  - ~~[ ] - MatrixAI/Polykey#161~~
- [x] 7. Add agent service `nodes` tests https://github.com/MatrixAI/js-polykey/issues/194#issuecomment-1032173720
  * [x] - `nodesChainDataGet`
  * [x] - `nodesClosestLocalNode`
  * [x] - `nodesHolePunchMessage`
- [x] 8. Changed `Cmd` to `EntryPoint` because PK only has 1 executable and makes `docker run ...` easier.
- [x] 9. The `--seed-nodes='<defaults>;...'` will now mean that any specified seed nodes overrides the defaults rather than the other way around
- [x] 10. If seed nodes contains the own `NodeID`, the own `NodeId` will be automatically filtered.
- [x] 11. Make `pk agent start` return status information such as `nodeId` and not just `recoveryCode`
- [x] 12. All PK commands can contact a remote agent (except `agent start` and the like), it needs to be made more obvious whether we are contacting the local agent or remote agent https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1040120023
  - This is done by making the required client connection parameters of node ID, client host and client port all or nothing when set. This means if they are partially set, this results in a new exception `ErrorCLIClientOptions`.
  - If they are all set, that means a connection to an agent not specified by the local status.
  - If none are set, this means relying on the local status.
- [x] 13. The seed node should be adding the details of any node connecting to it to its own `NodeGraph`: see issue MatrixAI/Polykey#344
    - currently the seed node does **not** add an entry in its `NodeGraph` for a connecting node
    - this is essential for the Kademlia implementation to start to function, and for the node's details to eventually propagate out into the wider Polykey network
    - see https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1042518638
- [x] 14. Investigate and fix encountered Polykey errors/problems during testnet testing https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1041190329:
   - ~~[ ] a. No output provided on `pk identities trust`? However, I understand that we've previously shied away from providing "success" output on every single one of our commands. It can be tricky to diagnose problems when this is the case though.~~ - moved this to https://github.com/MatrixAI/js-polykey/issues/334#issuecomment-1043779027
   - ~~[ ] b. `pk identities trust` seemingly not adding a node to the gestalt graph when it doesn't already exist in `GestaltGraph`/`NodeGraph` - see https://github.com/MatrixAI/js-polykey/pull/326#issuecomment-1042543612 and https://matrixai.slack.com/archives/CEAUUV5QX/p1645069875382019~~ - moved this to https://github.com/MatrixAI/js-polykey/issues/334#issuecomment-1043779027
   - [x] c. `SyntaxError: Unexpected end of data` on an "invalid" node ID when using `decodeNodeId` - rebased on master has new js-id 3.3.1 that catches syntax errors when decoding multibase encoded strings
- [x] 15. Refactor `NodeGraph` structure
  * [x] The new structure will use 3 nested sublevels to achieve MatrixAI/Polykey#244 
    - `buckets` sublevel will contain each bucket, where each bucket sublevel contains `NodeId` to `NodeData`
    - `meta` sublevel contains each bucket, where each bucket sublevel is a config structure, currently only `count`
    - `lastUpdated` or `index` sublevel that contains each bucket, and each bucket sublevel contains `lexi(lastUpdated)-NodeId` to `NodeId`, the key is a compound index, allowing us to efficiently acquire the most up-to-date or least up-to-date node entry
  * [x] New methods like `getNodes` and `getBuckets` will allow us to interrogate the state of the `NodeGraph` efficiently, by streaming data out of the `NodeGraph`, this will be important for debugging, and later analytics
  * [x] Integration of `DBTransaction` and new iterator feature of sublevels that allow us to maintain a snapshot of the DB at a point in time
  * [x] Reviewing kademlia trie structure and bringing ideas in - reviewed and discovered that trie structure properties are not necessary in our NodeGraph, our design is already efficient and optimal
- [x] 16. Optimise `getClosestNode` method based on new `NodeGraph` structure
- [x] 17. Relationship between `NodeGraph` and `NodeManager` should be tightly coupled due to the adding and removal policies of nodes, maintaining the bucket limit must involving pinging old nodes
- [x] 18. tests for changing root keys and connections.
  - Check how active connections work during a keychange.
  - Check if new connections use new TLS config
  - check if a renew/reset key event propigates to where it is needed.
-  [x] 19. cleaning up and small fixes to nodesUtils parser functions
-  [x] 20. adding force and ping flags to nodes add command


### Final checklist
<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

* [x] Domain specific tests
* [x] Full tests
* [x] Updated inline-comment documentation
* [x] Lint fixed
* [x] Squash and rebased
* [x] Sanity check the final build

joshuakarp committed:
feat: added `nodesGetAll` bin command and service handler

Retrieves all buckets from the NodeGraph

CMCDragonkai committed:
fix: general fixes

- added getBuckets test for distance and lastUpdated order
- resetting buckets work
- changing utility names

tegefaulkes committed:
feat: added `NodeGraph.getClosestNodes()`

Implemented `getClosestNodes()` and relevant tests in `NodeGraph.test.ts`.

Relates to #212

tegefaulkes committed:
fix: `NodeManager.setNode` Properly handles adding new node when bucket is full

Logic of adding nodes has been split between `NodeManager` and `NodeGraph`. The `NodeGraph.setNode` just handles adding a node to the bucket where the `NodeManager.setNode` contains the logic of when to add the node

Relates #359

tegefaulkes committed:
test: fixed `nodeGraph` `get all buckets` test

fix into 7cc74d059f94739591307d9543fa73f999d3345f

tegefaulkes committed:
feat: added `connectionEstablishedCallback` to `Proxy`

Added a callback to the `Proxy` that is called when a `ForwardConnection` or `ReverseConnection` is established and authenticated. It is called with the following connection information; `remoteNodeId`, `remoteHost`, `remotePort` and `type`. They type signifies if it was a forward or reverse connection. Note that this is only triggered by composed connections.

Added a test for if the callback was called when a `ReverseConnection` is established.

Relates #332
Relates #344

tegefaulkes committed:
feat: `Proxy` trigger adding nodes to `Nodegraph`

Added an event to the `EventBus` that is triggered by the `Proxy`'s `connectionEstablishedCallback`. this adds the node to the `NodeGraph`.

Related #344

tegefaulkes committed:
feat: added optional timeout timers to `NodeConnectionManager` methods

In some cases we want to specify how long we attempt to connect to a node on a per-connection basis.

Related #363

tegefaulkes committed:
refactor: updated implementation of `nodePing`

Added `nodePing` command to `NodeConnectionManager` and `NodeManager.nodePing` calls that now. the new `nodePing` command essentially attempts to create a connection to the target node. In this process we authenticate the connection and check that the nodeIds match. If no address is provided it will default to trying to find the node through kademlia.

Related #322

tegefaulkes committed:
feat: `NodeManager.setNode` authenticates the added node

`setNode` now authenticates the node you are trying to add. Added a flag for skipping this authentication as well as a timeout timer for the authentication. this is shared between authentication new node and the old node if the bucket is full.

Related #322

tegefaulkes committed:
syntax: general linting and fixes from api changes

tegefaulkes committed:
fix: squash into ping node changes.

Updated `NodeConnectionManager.pingNode` to just use the proxy connection.

#322

tegefaulkes committed:
fix: setNode now does not ping the new node

#322

tegefaulkes committed:
feat: setNode concurrently pings multiple nodes

`setNode` now pings 3 nodes concurrently, updating ones that respond and removing ones that don't. If there is room in the bucket afterwards then we add the new node.

#322

tegefaulkes committed:
feat: nodeManager is now startStop

#322

tegefaulkes committed:
feat: async queueing for setting nodes

`setNode` now has a `blocking` flag that defaults to false. If it encounters a full bucket when adding a node then it will add the operation to the queue and asynchronously trys a blocking `setNode` in the background. `setNode`s will only be added to the queue if the bucket was full.

#322

tegefaulkes committed:
feat: establishing a `NodeConnection` adds the node to the nodeGraph

`NodeConnectionManager` now takes `NodeGraph` in  the `nodeConnectionManager.start` method. It has to be part of the start method since they are co-dependent. `NodeConnectionManager` cals `NodeManager.setNode()` when a connection is established. This fulfills the condition of adding a node to the graph during a forward connection.

Fixed up tests that were failing in relation to the `NodeManager` `StartStop` conversion.

#322

tegefaulkes committed:
feat: implemented `NodeManager.refreshBucket()`

This method preforms the kademlia `refreshBucket` operation. It selects a random node within the bucket and preforms a search for that node. The process exchanges node information with any nodes it connects to.

#345

tegefaulkes committed:
feat: implemented no activity timers and queuing for `refreshBucket`

Added queuing for `refreshBucket`. This means that buckets will be refreshed one at a time sequentially. This is to avoid doing a lot of costly refreshing all at once.

Added no activity for buckets. If a bucket hasn't been touched for a while, 1 hour by default, it will add a refresh bucket operation to the queue. Timers are disabled for buckets already in the queue. Only 1 timer is used for all buckets since only one of them can have the shortest timer and that's all we really care about.

#345

tegefaulkes committed:
feat: refreshing buckets when entering network

`nodeConnectionManager.syncNodeGraph` now refreshes all buckets above the closest node as per the kademlia spec. This means adding a lot of buckets to the refresh bucket queue when an agent is started.

#345

tegefaulkes committed:
feat: abort controller support for `NodeManager.refreshBucket`

Added support to cancel out of a `refreshBucket` operation. This is to allow faster stopping of the `NodeManager` by aborting out of a slow `refreshBucket` operation. This has been implemented with the `AbortController`/`AbortSignal` API. This is not fully supported by Node14 so we're using the `node-abort-controller` to provide functionality for now.

#345

emmacasolin committed:
fix: `setNode` no longer adds node twice when `force` is true
#322

emmacasolin committed:
feat: generic `SetNodeQueue` class for queuing `setNode` operations
`NodeManager.setNode` and `NodeConnectionManager.syncNodeGraph` now utilise a single, shared queue to asynchronously add nodes to the node graph without blocking the main loop. These methods are both blocking by default but can be made non-blocking by setting the `block` parameter to false.
#322

emmacasolin committed:
fix: `syncNodeGraph` during agent startup is now non-blocking
#322

emmacasolin committed:
refactor: renamed `SetNodeQueue` to `Queue`

Renamed `queueStart` and `queuePush` since `Queue` is its own class now
Simplified some logic using the `promise` utility

tegefaulkes committed:
tests: general fixes for failing tests

This contains fixes for failing tests as well as fixes for tests failing to exit when finished.

tegefaulkes committed:
syntax: added `@typescript-eslint/await-thenable` linting rule

This checks if we await things that are not promises. This is not a problem per se, but we generally don't want to await random things.

tegefaulkes committed:
fix: updated `@types/node-forge` version and fixed `keysUtils.getRandomBytes`

tegefaulkes committed:
tests: added test to check if nodes are properly added to the seed nodes when entering the network

This tests for if the Seed node contains the new nodes when they are created. It also checks if the new nodes discover each other after being created.

Includes a change to `findNode`. It will no longer throw an error when failing to find the node. This will have to be thrown by the caller now. This was required by `refreshBucket` since it's very likely that we can't find the random node it is looking for.

tegefaulkes committed:
tests: Added agent service tests for `nodesChainDataGet`, `nodesClosestLocalNode` and `nodesHolePunchMessage`

tegefaulkes committed:
fix: post re-base fixes

DB and type changes using new transactions
Linting
Fixing timeouts

CMCDragonkai commented on src/utils/utils.ts:
```diff
@@ -154,14 +154,16 @@ function promisify<T>(f): (...args: any[]) => Promise<T> {
- * Deconstructed promise
- */
-function promise<T>(): {
+export type PromiseType<T> = {
```
1. The `export` should be at the very end.
2. This is a nominal type for the deconstructed promise, prefer calling this `PromiseDeconstructed`.

CMCDragonkai commented on src/utils/utils.ts:
```diff
@@ -220,6 +222,67 @@ function arrayZipWithPadding<T1, T2>(
+  return arr;
+}
+
+function bufferSplit(
```
This may not be necessary anymore if we are able to use `KeyPath` directly.

Primarily for composite keys where we wanted to have it sorted by a particular index, but still acquire the second part of the key.

But don't remove this function because it may be useful in the future.

CMCDragonkai replied:
Look at `lexicographic level iteration order` test in js-db to see the rules of iteration that the DB does now, then you can see if it applies to what we want to use for composite keys.

CMCDragonkai commented on src/nodes/utils.ts:
```diff
@@ -1,29 +1,75 @@
+import { bytes2BigInt, bufferSplit } from '../utils';
+import * as keysUtils from '../keys/utils';
+
+// FIXME:
```
Shoudl look for all the FIXMEs and resolve them. Also remove any spurious commentary.

tegefaulkes committed:
syntax: style fixes

Exports moved to end of the file as part of the `export {}` block
Cleaned up TODOs and FIXMEs
Fixed index and other incorrect imports
Type fixes in utils

tegefaulkes committed:
build: removing `node-abort-controller` polyfill

Abort controller functionality is included in node now.

tegefaulkes committed:
build: removing unneeded files

tegefaulkes committed:
fix: `NodeConnectionManager.syncNodeGraph` now pings nodes

Need to ensure validity of nodes by pinging them before adding them to the node graph.

#322

tegefaulkes committed:
fix: small fixes to nodes utils parsing functions

tegefaulkes committed:
tests: expanding tests for root keypair changing

Added some tests to check that a root keyPair change propagates properly. Also added tests for the change for existing and new node connections.

#317

tegefaulkes committed:
build: updating package-lock.json

tegefaulkes committed:
fix: removed unneeded timer from NodeConnection test
